packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1824,9 +1824,106 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
   return root;
 }
 
-export function warnIfNotCurrentlyBatchingInDev(fiber: Fiber): void {
+// in a test-like environment, we want to warn if dispatchAction() is
+// called outside of a TestUtils.act(...)/batchedUpdates/render call.
+// so we have a a step counter for when we descend/ascend from
+// actedUpdates() calls, and test on it for when to warn
+let actingUpdatesScopeDepth = 0;
+
+// we 'wait' for queued microtasks to resolve (after flushing effects)
+// using either setImmediate when available, or postMessage
+const nextTick =
+  typeof setImmediate !== 'undefined'
+    ? setImmediate
```

Comment:
What if `setImmediate` is mocked? [jest mocks it](https://github.com/facebook/jest/blob/master/packages/jest-fake-timers/src/jestFakeTimers.ts)

Code like this would pick up the mocked version:

```js
jest.useFakeTimers();
let ReactDOM = require('react-dom');
```

In that setup we wouldn't flush any work at the end of act until timers are ran. This is bad. We should make this just work.

Is this the reason you have to do weird things like call useRealTimers and call runAllTimers at random places?

Reply:
If we can somehow instead `require('timers').setImmediate` from the Node environment, we can get to the unmocked version.

```
let setImmediate;
try {
  let r = require; // trick packagers not to bundle this stuff.
  setImmediate = r('timers').setImmediate;
} catch (x) {
  // try one of the other choices
}
```

Reply:
> Is this the reason you have to do weird things like call useRealTimers and call runAllTimers at random places?

(notes for posterity's sake)

I need to `useRealTimers()` because we use fake timers by default, but OSS folks don't, so I need to explicitly disable them for our tests. 
There are 4 'scenarios' -
- fake timers, with transpiled environments (alá fb): in this setup, the async version of act is barely (never?) used, and peeps use `runAllTimers()` to 'advance' time. of note, the above codepath is never used. this is typical code when so - 
```jsx
act(() => {
  jest.runAllTimers();
  // advance time and capture 
  // updates and effects
});
``` 
- real timers, common js presets: closer to what OSS has. in this scenario, both the sync and async versions work as expected, no real surprises. you would see code like this -
```jsx
await act(async () => {
  doSomething()
  await waitForElement(() => getByTestId('greeting-text'))
}) 
```
- real timers, transpiled environments: similar to the previous one, both sync/async versions will work as expected. 
- fake timers, common js presets: this is the default setup we have in the react repo, and where tests with cascading promises would fail (it'll pass the first tick with `await act(async () =>   jest.runAllTimers();`, but thats not sufficient). your proposed hack fixes this! for browsers without MessageChannel, how do you feel about this hack? https://github.com/facebook/react/pull/14853/#discussion_r265993434

Reply:
> act is barely (never?) used, and peeps use runAllTimers() to 'advance' time. of note, the above codepath is never used

I'm confused about the expectation here. I think this is what we explicitly __don't__ want to have happen.

Best practice is to not call runAllTimers inside act because act only means that the event turns have happened, not that time has elapsed. An event turn can execute infinitely quickly.

For example I often want do:

```js
await act(() => {
  setState();
});
// expect on the first commit
await act(() => {
  runAllTimers();
});
// expect on subsequent commits that happened after timers set up in effects fired
```

Similarly sometimes you want to call timers without actually flushing any rendering work.

```js
await act(() => {
  setState();
  runAllTimers();
  // the set state above is still not flushed because it might have taken awhile to render
});
// expect on the commit
```

Is this also your expectation?

Reply:
I think we're on the same page? The examples you just gave should still work. I simply meant that *in an fb-like-environment*, there isn't much use (yet!) for the *async version of act*, since - 
```jsx
act(() => {
  runAllTimers();
});
```
will effectively do the same thing as -
```jsx 
await act(async () => {
  runAllTimers();
});
```
(which is to resolve queued promises and timers.)